java/ql/lib/semmle/code/java/regex/regex.qll

```diff
@@ -0,0 +1,933 @@
+import java
+private import RegexFlowConfigs
+
+// In all ranges handled by this library, `start` is inclusive and `end` is exclusive.
+/**
+ * A string literal that is used as a regular expression.
+ */
+abstract class RegexString extends StringLiteral {
+  /** Holds if a character set starts between `start` and `end`, including any negation character (`^`). */
+  private predicate charSetStart0(int start, int end) {
+    this.nonEscapedCharAt(start) = "[" and
+    (if this.getChar(start + 1) = "^" then end = start + 2 else end = start + 1)
+  }
+
+  /** Holds if the character at `pos` marks the end of a character class. */
+  private predicate charSetEnd0(int pos) {
+    this.nonEscapedCharAt(pos) = "]" and
+    /* special case: `[]]` and `[^]]` are valid char classes.  */
+    not this.charSetStart0(_, pos)
+  }
+
+  /**
+   * Gets the nesting depth of character classes at position `pos`
+   */
+  private int charSetDepth(int pos) {
+    exists(this.getChar(pos)) and
+    result =
+      max(int j |
+        j = 0 or
+        j =
+          count(int i | i < pos and this.charSetStart0(i, _)) -
+            count(int i | i < pos and this.charSetEnd0(i))
+      )
```

Comment:
This seems inefficient, and given how common character classes are, we should probably fix this. Replace by applying `rank` to charSet starts and ends, then do a linear recursion over those rank-indices to calculate the depth, since you only need the depth info at those positions anyway.